### PR TITLE
RPC password authentication

### PR DESCRIPTION
**Pull request by rorp**

### CODE REVIEW

**Review by Christewart — COMMENTED** [review]
Could you add a negative unit test to make sure bad passwords fail?

File: app/cli/src/main/scala/org/bitcoins/cli/ConsoleCli.scala [hunk 1]
```diff
@@ -2011,6 +2014,8 @@ object ConsoleCli {
         sttp
           .post(uri"http://$host:${config.rpcPort}/")
           .contentType("application/json")
+          .auth
+          .basic("bitcoins", config.rpcPassword)
```
Comment (benthecarman) [inline app/cli/src/main/scala/org/bitcoins/cli/ConsoleCli.scala:2018]: why not let the rpc user be set too like most other applications?

**Comment by Christewart** [comment]
Compile failing https://github.com/bitcoin-s/bitcoin-s/runs/4708619274?check_suite_focus=true#step:5:207

**Comment by benthecarman** [comment]
I copied this is another project. It currently has a problem where if you don't send the correct authentication it doesn't send an error back, just an empty result and error.

File: app/cli/src/main/scala/org/bitcoins/cli/ConsoleCli.scala (hunk 1 above)
Comment (rorp) [reply app/cli/src/main/scala/org/bitcoins/cli/ConsoleCli.scala:2018]: Because there's only one user. It doest's matter what's its name.

**Comment by rorp** [comment]
> I copied this is another project. It currently has a problem where if you don't send the correct authentication it doesn't send an error back, just an empty result and error.

What do you mean? It responds as RFC7235 requires:

```shell
$ curl -i --data-binary '{"jsonrpc": "1.0", "id": "curltest", "method": "getinfo", "params": []}' -H "Content-Type: application/json" http://127.0.0.1:9999/
HTTP/1.1 401 Unauthorized
WWW-Authenticate: Basic realm="auth",charset=UTF-8
Server: akka-http/10.2.7
Date: Mon, 10 Jan 2022 19:42:56 GMT
Content-Type: text/plain; charset=UTF-8
Content-Length: 77

The resource requires authentication, which was not supplied with the request
```

```shell
$ curl -i -u user:pwd --data-binary '{"jsonrpc": "1.0", "id": "curltest", "method": "getinfo", "params": []}' -H "Content-Type: application/json" http://127.0.0.1:9999/
HTTP/1.1 401 Unauthorized
WWW-Authenticate: Basic realm="auth",charset=UTF-8
Server: akka-http/10.2.7
Date: Mon, 10 Jan 2022 19:49:03 GMT
Content-Type: text/plain; charset=UTF-8
Content-Length: 38

The supplied authentication is invalid
```

File: app/server-routes/src/main/scala/org/bitcoins/server/routes/Server.scala [hunk 2]
```diff
@@ -29,6 +29,15 @@ case class Server(
 
   import system.dispatcher
 
+  if (rpcPassword.isEmpty) {
+    if (rpchost == "localhost" || rpchost == "127.0.0.1") {
```
Comment (Christewart) [inline app/server-routes/src/main/scala/org/bitcoins/server/routes/Server.scala:34]: This seems like a really good idea. Best of both worlds IMO :+1:

File: app/cli/src/main/scala/org/bitcoins/cli/ConsoleCli.scala (hunk 1 above)
Comment (Christewart) [reply app/cli/src/main/scala/org/bitcoins/cli/ConsoleCli.scala:2018]: > why not let the rpc user be set too like most other applications?

We are copying what eclair does for their backend. I'm not sure how we can correctly map users to access to backend services (for instance, does Alice have access to "bob-wallet") and vice versa

File: app/server-routes/src/main/scala/org/bitcoins/server/routes/Server.scala [hunk 3]
```diff
@@ -65,19 +66,29 @@ case class Server(
     }
   }
 
+  def authenticator(credentials: Credentials): Option[Done] =
```
Comment (Christewart) [inline app/server-routes/src/main/scala/org/bitcoins/server/routes/Server.scala]: Consider making `private`

File: app/server-routes/src/main/scala/org/bitcoins/server/routes/Server.scala [hunk 4]
```diff
@@ -119,8 +130,10 @@ case class Server(
   private val eventsRoute = "events"
 
   private def wsRoutes: Route = {
-    path(eventsRoute) {
-      Directives.handleWebSocketMessages(wsHandler)
+    authenticateBasic("auth", authenticator) { _ =>
```
Comment (Christewart) [inline app/server-routes/src/main/scala/org/bitcoins/server/routes/Server.scala:147]: Whats the benefit of adding auth on the websocket? Currently we only write things to the websocket and do not support requests sent to us over the websocket.

**Review by Christewart — CHANGES_REQUESTED** [review]
I see a unit test for websocket authorization but not the rpc

File: app/server-routes/src/main/scala/org/bitcoins/server/routes/Server.scala (hunk 4 above)
Comment (rorp) [reply app/server-routes/src/main/scala/org/bitcoins/server/routes/Server.scala:147]: Same benefit as for the HTTP server - it adds access control.

**Review by Christewart — APPROVED** [review]

**Pull request merged**